dnf-behave-tests/features/comps-group-upgrade.feature

```diff
@@ -0,0 +1,100 @@
+  Given I use repository "comps-group-upgrade-1"
+    And I successfully execute dnf with args "install dummy"
+    And I successfully execute dnf with args "group mark install testgroup"
+
+
+Scenario: Upgrade group when there are no new packages - nothing is installed
+   When I execute dnf with args "group upgrade testgroup"
+   Then the exit code is 0
+    And Transaction is following
+        | Action        | Package                            |
+        | group-upgrade | testgroup                          |
+
+
+Scenario: Upgrade group when there are new packages - install new packages
+  Given I drop repository "comps-group-upgrade-1"
+    And I use repository "comps-group-upgrade-2"
+   When I execute dnf with args "group upgrade testgroup"
+   Then the exit code is 0
+    And Transaction is following
+        | Action        | Package                            |
+        | install-group | mandatory-2-0:1.0-1.x86_64         |
+        | install-group | default-2-0:1.0-1.x86_64           |
+        | install-group | conditional-true-2-0:1.0-1.x86_64  |
+        | group-upgrade | testgroup                          |
+
+
+Scenario: Upgrade group when there are both old and new packages - install only new packages
+      # I don't drop repository comps-group-upgrade-1, so the comps are merged
+  Given I use repository "comps-group-upgrade-2"
+   When I execute dnf with args "group upgrade testgroup"
+   Then the exit code is 0
+    And Transaction is following
+        | Action        | Package                            |
+        | install-group | mandatory-2-0:1.0-1.x86_64         |
+        | install-group | default-2-0:1.0-1.x86_64           |
+        | install-group | conditional-true-2-0:1.0-1.x86_64  |
+        | group-upgrade | testgroup                          |
+
+
+Scenario: Upgrade group to new metadata and back - always install new packages
+  Given I drop repository "comps-group-upgrade-1"
+    And I use repository "comps-group-upgrade-2"
+   When I execute dnf with args "group upgrade testgroup"
+   Then the exit code is 0
+    And Transaction is following
+        | Action        | Package                            |
+        | install-group | mandatory-2-0:1.0-1.x86_64         |
+        | install-group | default-2-0:1.0-1.x86_64           |
+        | install-group | conditional-true-2-0:1.0-1.x86_64  |
+        | group-upgrade | testgroup                          |
+  Given I drop repository "comps-group-upgrade-2"
+    And I use repository "comps-group-upgrade-1"
+   When I execute dnf with args "group upgrade testgroup"
+   Then the exit code is 0
+    And Transaction is following
+        | Action        | Package                            |
+        | install-group | mandatory-0:1.0-1.x86_64           |
+        | install-group | default-0:1.0-1.x86_64             |
+        | install-group | conditional-true-0:1.0-1.x86_64    |
+        | group-upgrade | testgroup                          |
```

Comment:
Hi, did I miss something or do the tests actually not test upgrading packages to a new version? You've mark-installed in `Background` and here it actually installs the missing packages, but `dnf group upgrade` should also upgrade already installed packages, right?

Reply:
So far, I was only interested in picking the right packages to install when doing group upgrade (e.g. not installing reinstalled packages or packages that were already part of the group but were not installed) to determine if we can do group upgrade during system upgrade (see https://github.com/rpm-software-management/dnf-plugins-extras/pull/183). In the repositories that I defined, there are no packages in multiple versions, so nothing gets upgraded in these scenarios. But it's a good point, I should also add test the actual package upgrade.

Reply:
Hm, the naming of packages is probably confusing, because there is e.g. mandatory and mandatory-2 that are different packages. Maybe it would be better to rename them to mandatory-A and mandatory-B.

Reply:
I added test for upgrading packages during group upgrade and renamed the packages. Hope it's better readable now.